### ISSUE TITLE
`&usize::<=` is much slower than `usize::<=`

### ISSUE THREAD

**Issue by hashworks** [rust-lang/rust#105259]
While solving [AoC Day 4](https://adventofcode.com/2022/day/4) I ran into the following performance difference:

```rust
fn main() {
    let input = vec![
        (2, 4, 6, 8),
        (2, 3, 4, 5),
        (5, 7, 7, 9),
        (2, 8, 3, 7),
        (6, 6, 4, 6),
        (2, 6, 4, 8), // .. 500000000 lines of random data, read from disk with real code (~12GB)
    ];

    // 1761ms on my machine
    let _variant_a_result = variant_a(&input);

    //  656ms on my machine
    let _variant_b_result = variant_b(&input);
}

pub fn variant_a(input: &[(usize, usize, usize, usize)]) -> usize {
    input
        .iter()
        .filter(|(a, b, c, d)| a <= c && d <= b || c <= a && b <= d)
        .count()
}

pub fn variant_b(input: &[(usize, usize, usize, usize)]) -> usize {
    input
        .iter()
        .filter(|&&(a, b, c, d)| a <= c && d <= b || c <= a && b <= d)
        .count()
}
```

As you can see, using `|&&(a, b, c, d)|` on the same closure is 2.6 times faster than `|(a, b, c, d)|`.
Viewing the compiled code makes this clear: https://rust.godbolt.org/z/oaeaGGPK7

Is this a compiler behavior a developer should know / expect?

### Meta

Thanks to @Arnavion for pointing this out.

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (c090c6880 2022-12-01)
```

**Comment by leonardo-m** [comment]
Unrelated: adding some parentesys could improve the clarity of this code:

    .filter(|(a, b, c, d)| (a <= c && d <= b) || (c <= a && b <= d))

**Comment by workingjubilee** [comment]
> Is this a compiler behavior a developer should know / expect?

Strictly answering this question: You are working with `&T` and then calling `.iter()`, which creates nested indirections. Yes, the cost of an indirection, especailly nested indirections, is indeed probably something you should be aware of when you are working with plain, "by-value" data, and thus you should probably try to avoid it.

Also your signature should probably be
```rust
fn(input: &[(u64, u64, u64, u64)]) -> usize
```
But that doesn't make a difference to codegen.

This is what motivated https://github.com/rust-lang/rust/issues/44762

**Comment by jruderman** [comment]
Why is `variant_a` valid, and why does it create a nested indirection? I sorta understand the non-tuple case described in [an old reddit thread](https://www.reddit.com/r/rust/comments/41nyha/help_me_understand_references_in_iterator_filters/), but I don't understand how you can destructure a tuple and still have the wrong level of indirection.

**Comment by leonardo-m** [comment]
Both versions are idiomatic. And I think Clippy doesn't fire on them.

**Comment by workingjubilee** [comment]
Yes, this code is not unusual and ideally we'd compile it optimally in all cases, but creating indirections is Known to sometimes affect performance.

> but I don't understand how you can destructure a tuple and still have the wrong level of indirection.

If you destructure some `&tuple` into `(A, B)`, you will get `(&A, &B)`. This happens due to... you know, I forget exactly which, at this point.

**Comment by jruderman** [comment]
I see. So `variant_a` binds `a` etc as references, and the comparisons are doing auto-derefs equivalent to:

```rust
pub fn variant_a_2(input: &[(usize, usize, usize, usize)]) -> usize {
    input
        .iter()
        .filter(|(a, b, c, d)| *a <= *c && *d <= *b || *c <= *a && *b <= *d)
        .count()
}
```

It only seems to keep one level of references, though.

**Comment by workingjubilee** [comment]
Indeed. The nesting is only somewhat relevant because it requires more work to remove both levels. And I believe it is hitting this implementation:
```rust
impl<A, B> PartialOrd<&B> for &A where
    A: PartialOrd<B> + ?Sized,
    B: ?Sized,
```

**Comment by estebank** [comment]
I implemented a lint in https://github.com/rust-lang/rust/pull/108372, but that lint might be more appropriate in clippy. Having said that, I wanted to see what the impact on crater would be.

**Comment by scottmcm** [comment]
This is at least highly related to https://github.com/rust-lang/rust/issues/106269#issuecomment-1374158580 and #103024

The *copying* isn't actually necessary to fix it.  All LLVM needs to know is that it's a good idea to pre-read them, so this (variant C) *also* fixes it: <https://rust.godbolt.org/z/W4hqEjj6c>
```rust
pub fn variant_c(input: &[(usize, usize, usize, usize)]) -> usize {
    input
        .iter()
        .filter(|(a, b, c, d)| ((d <= b) & (a <= c)) | ((c <= a) & (b <= d)))
        .count()
}
```

By using non-short-circuiting operations LLVM doesn't need to worry about "is it ok to read that?" because it'll have to read everything.  And then the right thing happens.

To further emphasize this, note that this (variant D) does **not** fix the problem, despite copying-via-dereference all the values, as the short-circuiting logic still means it's not "obviously" reading everything: <https://rust.godbolt.org/z/4zPfMejcP>
```rust
pub fn variant_d(input: &[(usize, usize, usize, usize)]) -> usize {
    input
        .iter()
        .filter(|(a, b, c, d)| *a <= *c && *d <= *b || *c <= *a && *b <= *d)
        .count()
}
```

I wonder if it would fix it if optimized builds always did a copy-read of all `&`s to small copy types as soon as they were initialized, and let LLVM just optimize them away if they aren't actually CSE'd into something useful...

**Comment by saethlin** [comment]
> I wonder if it would fix it if optimized builds always did a copy-read of all `&`s to small copy types as soon as they were initialized

I've considered doing this in a MIR optimization. Not sure I have the time/patience to attempt it right now, but the transformation shouldn't be wildly complicated: Find function that have `&T` arguments where `T: Freeze`, insert new locals which are great derefs of the references as the first statements, replace derefs of the parameters with the new locals.

As far as I know this is exactly the transformation that `dereferenceable` authorizes. I'm sure such a transformation couldn't be enabled by default, but it would be interesting to try out in a branch.

**Comment by the8472** [comment]
I've been playing around with this and it requires some enhanced persuasion to get LLVM to agree that all the values in a chunk-reference are read. After that things vectorize properly.

https://rust.godbolt.org/z/YxPEbjf1j
